enhancements/network/vrf-cni.md

```diff
@@ -0,0 +1,161 @@
+- [x] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+To allow isolation, and make it possible to have overlaps between the CIDRs of secondary networks (and between them and the pod's address space), we want to introduce a VRF meta cni plugin. Taking advantage of the chaining mechanism, it can be used to assign a secondary interface to a custom [VRF](https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/8/html/configuring_and_managing_networking/configuring-virtual-routing-and-forwarding-vrf_configuring-and-managing-networking).
+
+## Motivation
+
+Telco customers maintain strict separation of multiple networks for different services or functionalities. When a device connects to multiple networks, the practice is to maintain strict separation of these networks.
+
+Moreover, the multiple networks a pod is connected to may have overlapping CIDRs, and those CIDRs can also overlap with the sdn’s CIDR. To maintain this kind of isolation, they typically use VRFs. Currently this is done by granting the containers the NET_ADMIN capabilities.
+
+### Goals
+
+Being able to assign a secondary interface to a particular vrf, and eventually having more secondary interfaces of the same pod assigned to the same VRF.
+
+### Non-Goals
+
+Assigning host level interfaces to VRFs.
+
+## Proposal
+
+The proposal is to introduce an additional CNI plugin that takes the vrf name as an input, and can be chained to the “main” cni used for the secondary interface, and that has the effect of creating the requested vrf (if it does not exist) and to add the interface to that vrf.
+
+In this way, the network interface returned by any CNI can be isolated in its own VRF, satisfying the above requirements without the need of additional capabilities.
+
+### User Stories [optional]
+
+#### Story 1
+
+As a user, I want to be able to optionally specify a VRF which the secondary nic is attached to.
+
+### Implementation Details/Notes/Constraints
+
+By adding several networks with the same `vrf` name, the CNI will ensure that they are all added to the same vrf, creating it only the first time.
+
+A `NetworkAttachmentDefinition` for MacVlan would look like:
+
+```yaml
+apiVersion: "k8s.cni.cncf.io/v1"
+kind: NetworkAttachmentDefinition
+metadata:
+  name: macvlan-vrf-conf
+spec:
+  config: '{
+              "cniVersion": "0.3.1",
+              "plugins": [
+                {
+                  "type": "macvlan",
+                  "mode": "bridge",
+                  "ipam": {
+                    "type": "host-local",
+                    "subnet": "192.168.1.0/24",
+                    "rangeStart": "192.168.1.200",
+                    "rangeEnd": "192.168.1.216",
+                  }
+                },
+                {
+                  "type": "vrf",
+                  "vrfname": "blue",
+                  "routingtable": 1001
```

Comment:
Can we add different vrfname with same routingtable  in a container?

Example:
 - vrf: foobar, routingtable: 1001
 - vrf: foobar, routingtable: 1002

In this case, user expects that different vrf between (1st) foobar and (2nd) foobar...

Reply:
Good point! Let me try locally, my guessing is that a single vrf is associated to a single routing table because the association between vrf and routing table happens at vrf creation time.

In this case (device1 asks `foobar - 1001` / device2 asks `foobar - 1002`) we need to return an error since the vrf is already there but it's using a different routing table. The other approach is hiding the routing table inside the cni, as it was before, and pick a new one whenever a new vrf is created.

Reply:
Yep, that can't be done. We either need to return an error if the same vrf is added with a different routing table, or we must hide the routing table and generate the id in the CNI. 
Both options are fine, I think the previous one (with out the routing table id exposed) is more slick.

Reply:
I think table id can be hidden as long as vrf name is unique inside pod namespace.
How does user specify additional VRF routes to the associated table? is it one of the VRF CNI arguments?

Reply:
What vrf interface information shall be presented to application, and how? (for example, there are multiple vrfs in the pod, with each associated with one routing table)

Reply:
> I think table id can be hidden as long as vrf name is unique inside pod namespace.
> How does user specify additional VRF routes to the associated table? is it one of the VRF CNI arguments?

We have two options here I think:
- hiding the routing table inside the CNI implementation, and assume that each vrf belongs to a different routing table (which seems reasonable to me). This has the burden of finding a different id each time, but since the number of vrf will be limited inside a pod, I think that can be done

- letting the user specify the routing table associated to the vrf (as you do using `ip`), as per this last change here https://github.com/openshift/enhancements/pull/405/files/114b61cd4b1c6e975220156c435c011af9dc931f#diff-603b0353ba1a6867a490e42ab657f6dcR81
In this case we must ensure that two interfaces added to the same vrf specify the same routing table. It feels a bit clunky to me, I think I like the former most.

Reply:
> What vrf interface information shall be presented to application, and how? (for example, there are multiple vrfs in the pod, with each associated with one routing table)

This is a good point. I think the application will care about what device belongs to which VRF, assuming the vrf name is meaningful. This will allow it to bind to the right device when it wants to listen / send from a particular "slice".
We can inject an env variable with such association. WDYT?

Reply:
>     * hiding the routing table inside the CNI implementation, and assume that each vrf belongs to a different routing table (which seems reasonable to me). This has the burden of finding a different id each time, but since the number of vrf will be limited inside a pod, I think that can be done
> 
+1 on the first (above) option, I think we have 1:1 mapping between vrf name and route table. 
If there will be need to specify route table id, it can be added later per request (as a CNI argument).

Reply:
> This is a good point. I think the application will care about what device belongs to which VRF, assuming the vrf name is meaningful. This will allow it to bind to the right device when it wants to listen / send from a particular "slice".
> We can inject an env variable with such association. WDYT?

how will the env variable be injected, could you please elaborate?

Reply:
> How does user specify additional VRF routes to the associated table? is it one of the VRF CNI arguments?

will the route rules be one of the CNI argument?

Reply:
> will the route rules be one of the CNI argument?

Sorry I missed the question. I would say not, and leave that route configuration (and ip configuration) to the CNIs that get called before this gets invoked. From [here](https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/8/html/configuring_and_managing_networking/configuring-virtual-routing-and-forwarding-vrf_configuring-and-managing-networking) "Addresses and routes previously attached to the enslaved device will be moved inside the VRF domain."

Reply:
From CNI plugin user's point of view, we should support vrfname and routing table id because sometimes containers want to use some specific routing table id for specific VRF. VRF plugin just check that both, vrf name and routing table id, are unique in the node or not. How about that?

Reply:
> > This is a good point. I think the application will care about what device belongs to which VRF, assuming the vrf name is meaningful. This will allow it to bind to the right device when it wants to listen / send from a particular "slice".
> > We can inject an env variable with such association. WDYT?
> 
> how will the env variable be injected, could you please elaborate?

Ignore this, I need to do some research because it does not look a trivial thing to do from a CNI

Reply:
> From CNI plugin user's point of view, we should support vrfname and routing table id because sometimes containers want to use some specific routing table id for specific VRF. VRF plugin just check that both, vrf name and routing table id, are unique in the node or not. How about that?

The VRFs (and their tables) are namespaced, you can have two pods on the same node, create a vrf in each pod with the same name but different table ids (I just tried). I don't see much value in adding a specific tableid, but I may be missing something.

Reply:
> What vrf interface information shall be presented to application, and how? (for example, there are multiple vrfs in the pod, with each associated with one routing table)

Getting back on this, according to https://www.kernel.org/doc/Documentation/networking/vrf.txt it looks like a vrf aware application binds directly to the vrf interface. I assume it will accept the name of the interface as a parameter (or hardcoded!), as it has to know the device it binds to. 
So, when creating the vrf the user can give it a name that is known by the application in a way or another.

Reply:
> > What vrf interface information shall be presented to application, and how? (for example, there are multiple vrfs in the pod, with each associated with one routing table)
> 
> Getting back on this, according to https://www.kernel.org/doc/Documentation/networking/vrf.txt it looks like a vrf aware application binds directly to the vrf interface. I assume it will accept the name of the interface as a parameter (or hardcoded!), as it has to know the device it binds to.
> So, when creating the vrf the user can give it a name that is known by the application in a way or another.

ok, so application will accept the name (perhaps as an user specified environment variable in pod spec) to choose the right vrf device.

Reply:
> > will the route rules be one of the CNI argument?
> 
> Sorry I missed the question. I would say not, and leave that route configuration (and ip configuration) to the CNIs that get called before this gets invoked. From [here](https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/8/html/configuring_and_managing_networking/configuring-virtual-routing-and-forwarding-vrf_configuring-and-managing-networking) "Addresses and routes previously attached to the enslaved device will be moved inside the VRF domain."

It's also mentioned in this [doc](https://github.com/openshift/enhancements/pull/405) that `Any additional routes depending on the enslaved device are dropped and will need to be reinserted to the VRF FIB table following the enslavement.` but I don't if this is a valid use case.

Reply:
> > > will the route rules be one of the CNI argument?
> > 
> > 
> > Sorry I missed the question. I would say not, and leave that route configuration (and ip configuration) to the CNIs that get called before this gets invoked. From [here](https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/8/html/configuring_and_managing_networking/configuring-virtual-routing-and-forwarding-vrf_configuring-and-managing-networking) "Addresses and routes previously attached to the enslaved device will be moved inside the VRF domain."
> 
> It's also mentioned in this [doc](https://github.com/openshift/enhancements/pull/405) that `Any additional routes depending on the enslaved device are dropped and will need to be reinserted to the VRF FIB table following the enslavement.` but I don't know if this is a valid use case.

Yep. I am not sure either. I think that could be added afterwards as we start gathering feedbacks. Does it make sense?

Reply:
makes sense